pkg/parse/listener_impl.go

```diff
@@ -82,10 +86,247 @@ func NewTreeShapeListener() *TreeShapeListener {
+	}
+	return nil
+}
+
+func newLinterRecords() *linterRecords {
+	return &linterRecords{
+		make(map[string]*graph),
+		newAppEndpointGraph(),
+	}
+}
+
+func (s *TreeShapeListener) getApps() *graph {
+	appName := s.getFullAppName()
+	if apps, exists := s.linter.apps[strings.ToLower(appName)]; exists {
+		return apps
+	}
+	rec := newAppEndpointGraph()
+	s.linter.apps[strings.ToLower(appName)] = rec
+	return rec
+}
+
+func (s *TreeShapeListener) createLocation(lineNum, colNum int) string {
+	return fmt.Sprintf("%s:%d:%d", s.sc.filename, lineNum, colNum)
+}
+
+func (s *TreeShapeListener) recordApp(location string) {
+	if !s.lintMode {
+		return
+	}
+	appName := s.getFullAppName()
+	if err := s.getApps().recordApp(appName, location); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+func (s *TreeShapeListener) getFullAppName() string {
+	// handle subpackage
+	return strings.Join(s.currentApp().Name.Part, "::")
+}
+
+func (s *TreeShapeListener) recordEndpoint(endpoint, location string) {
+	if !s.lintMode {
+		return
+	}
+	appName := s.getFullAppName()
+	if err := s.getApps().recordEndpoint(appName, endpoint, location); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+func (s *TreeShapeListener) recordMethod(endpoint, method, location string) {
+	if !s.lintMode {
+		return
+	}
+
+	appName := s.getFullAppName()
+	s.getApps().recordEndpoint(appName, endpoint, location) //nolint:errcheck
+	if err := s.getApps().recordMethod(appName, endpoint, method, location); err != nil {
+		logrus.Warn(err)
+	}
+}
+
+func (s *TreeShapeListener) recordCall(appName, endpoint, method, location string) {
+	if !s.lintMode {
+		return
+	}
+	if err := s.linter.calls.recordAsCall(appName, endpoint, method, location); err != nil {
+		logrus.Warn(err)
+	}
+}
+
+func (s *TreeShapeListener) lintEndpoint() {
+	appNotExistLog := func(location, appName, call string) {
+		logrus.Warnf("lint %s: Application '%s' does not exist for call '%s'", location, appName, call)
+	}
+	lint := func(appName, endpoint, method, call string, locations map[string]bool) {
+		for location := range locations {
+			apps, exists := s.linter.apps[strings.ToLower(appName)]
+			if !exists {
+				appNotExistLog(location, appName, call)
+				continue
+			}
+			if app, exists := (*apps)[appName]; exists {
+				if endpoints, exists := (*app.rec)[endpoint]; exists {
+					// if method is empty string, it is linting simple endpoint
+					if method != "" {
+						if _, exists = (*endpoints.rec)[method]; exists {
+							continue
+						}
+						logrus.Warnf("lint %s: Method '%s' does not exist for call '%s'", location, method, call)
+						continue
+					}
+				}
+				logrus.Warnf("lint %s: Endpoint '%s' does not exist for call '%s'", location, endpoint, call)
+				continue
+			}
+			appNotExistLog(location, appName, call)
+		}
+	}
+
+	for appName, appData := range *s.linter.calls {
+		for endpoint, endpointData := range *appData.rec {
+			// if method maps == nil, it is not REST endpoint
+			if endpointData.rec == nil {
```

Comment:
Nitpick: Could potentially pull this out into a function for readability, but it's pretty readable as is, so just presenting this as an option
```
if isNotRESTEndpoint {
}

func isNotRESTEndpoint(rec Record) bool {
    return endpointData.rec == nil
}
```

Reply:
it's meant to be only used at a certain level of the graph, I think it would create confusion if I make a function for it.